### PR TITLE
New VSphere periodic job and workflow competence

### PR DESCRIPTION
**Pull request by michaellevy101**
Adding new vsphere periodic job
Verify cluster installation
Gathering installation logs
Teardown vsphere cluster
Sharing inventory file between steps

### CODE REVIEW

**Comment by osherdp** [comment]
@michaellevy101 is it ready for review?

**Comment by michaellevy101** [comment]
> @michaellevy101 is it ready for review?

basically yes, ignore the debug prints and sleep instruction

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        KUBECONFIG: "{{ item.path }}"
+        LOGS_DEST: "/tmp/artifacts/new_cluster_{{ item.path | basename }}"
+      args:
+        chdir: /home/assisted
+      loop: "{{ kubeconfig_files.files }}"
+      loop_control:
+        label: "{{ item.path }}"
+EOF
+
+sleep 45m
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh:119]: ...
:)

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        ls -ltr /var/lib/swtpm-localca/ >> {{ LOGS_DIR }}/libvirt-qemu/ls-swtpm-localca.txt
+    - name: Check minikube kubeconfig file existence
+      stat:
+        path: /root/.kube/config
+      register: kubeconfig
+    - name: Extract assisted service logs
+      ansible.builtin.shell: |
+        make download_service_logs
+      args:
+        chdir: /home/assisted
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh:51]:
```suggestion
      make:
        chdir: /home/assisted
        target: download_service_logs
```

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      environment:
+        KUBECONFIG: "/root/.kube/config"
+      when: kubeconfig.stat.exists
+    - name: Extract capi logs
+      ansible.builtin.shell: |
+        make download_capi_logs
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh:57]: same as above

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+      args:
+        chdir: /home/assisted
+    - name: Find all log files
+      find:
+        paths: /home/assisted/
+        patterns: "*.log"
+        recurse: true
+      register: log_files
+    - name: Print log file names
+      debug:
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh:81]: is it just for debugging?
also for those other printings

**Comment by osherdp** [comment]
/cc @adriengentil 
FYI

**Comment by osherdp** [comment]
> > @michaellevy101 is it ready for review?
> 
> basically yes, ignore the debug prints and sleep instruction

ack
so ignore those comments that I already made :)

File: ci-operator/config/openshift/assisted-test-infra/openshift-assisted-test-infra-master.yaml [hunk 5]
```diff
@@ -517,7 +517,8 @@ tests:
     cluster_profile: vsphere
     env:
       ASSISTED_CONFIG: |
-        WORKER_COUNT=0
+        NUM_WORKERS=0
+        MAKEFILE_TARGET=test
```
[the remark this answers is not in the document]
Comment (michaellevy101) [reply ci-operator/config/openshift/assisted-test-infra/openshift-assisted-test-infra-master.yaml:521]: yes, complex logs for parallel jobs. don't see the benefit

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-ref.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
+ref:
+  as: assisted-common-gather
+  best_effort: true
+  timeout: 120m
+  optional_on_success: true
```
[the remark this answers is not in the document]
Comment (michaellevy101) [reply ci-operator/step-registry/assisted/common/gather/assisted-common-gather-ref.yaml:5]: can you please elaborate?

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh (hunk 1 above)
Comment (michaellevy101) [reply ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh:119]: still debugging:)

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh (hunk 4 above)
Comment (michaellevy101) [reply ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh:81]: it's for debugging but i'm considering keeping it. WDYT?

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-ref.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (osherdp) [reply ci-operator/step-registry/assisted/common/gather/assisted-common-gather-ref.yaml:5]: iiuc, it means if pre+test steps are passing, then we don't apply gather step
If you have a task to add/fix artifacts gathering, then it means you also need to add a logic to fail installation (and revert it afterwards).
Another case is when there's a false-positive in testing logic, that produces no error for failing states (then you cannot debug it without modifying the code)

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh (hunk 4 above)
Comment (osherdp) [reply ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh:79]: Sure no problem

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-ref.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (michaellevy101) [reply ci-operator/step-registry/assisted/common/gather/assisted-common-gather-ref.yaml:5]: thanks but the existing implementation is the same
https://github.com/openshift/release/blob/29112064321b0bc26611ef7aba5a925fbe45ee18/ci-operator/step-registry/assisted/baremetal/gather/assisted-baremetal-gather-ref.yaml

[the remark this answers is not in the document]
Comment (michaellevy101) [reply ci-operator/step-registry/assisted/common/gather/assisted-common-gather-ref.yaml:5]: ok. let give it a try. will change it to be mandatory

File: ci-operator/jobs/openshift/assisted-test-infra/openshift-assisted-test-infra-master-periodics.yaml [hunk 7]
```diff
@@ -1649,6 +1649,76 @@ periodics:
...
+  extra_refs:
+  - base_ref: master
+    org: openshift
+    repo: assisted-test-infra
+  labels:
+    ci-operator.openshift.io/cloud: vsphere
+    ci-operator.openshift.io/cloud-cluster-profile: vsphere
+    ci.openshift.io/generator: prowgen
+    pj-rehearse.openshift.io/can-be-rehearsed: "true"
+  name: periodic-ci-openshift-assisted-test-infra-master-e2e-vsphere-assisted-periodic
```
Comment (osherdp) [inline ci-operator/jobs/openshift/assisted-test-infra/openshift-assisted-test-infra-master-periodics.yaml:1665]: Please add slack notifications

File: ci-operator/step-registry/assisted/vsphere/teardown/cluster/assisted-vsphere-teardown-cluster-commands.sh [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+echo "************ vsphere assisted test-infra teardown cluster command ************"
+
+# Get packet | vsphere configuration
+# shellcheck source=/dev/null
+set +e
+source "${SHARED_DIR}/packet-conf.sh"
+source "${SHARED_DIR}/ci-machine-config.sh"
+set -e
+
+sleep 30m
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/vsphere/teardown/cluster/assisted-vsphere-teardown-cluster-commands.sh:16]: Seems like a waste of time, no?

File: ci-operator/step-registry/assisted/vsphere/teardown/cluster/assisted-vsphere-teardown-cluster-ref.yaml [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+      memory: 500Mi
+  credentials:
+  - namespace: test-credentials
+    name: vsphere-template-ssh-keys
+    mount_path: /var/run/vault/sshkeys
+  - namespace: test-credentials
+    name: edge-group-assisted-installer-bot
+    mount_path: /var/run/assisted-installer-bot
+  documentation: |-
+    Destory the Openshift cluster
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/vsphere/teardown/cluster/assisted-vsphere-teardown-cluster-ref.yaml:17]: Typo

File: ci-operator/step-registry/assisted/vsphere/teardown/machine/assisted-vsphere-teardown-machine-commands.sh [hunk 10]
```diff
@@ -12,9 +12,14 @@ if [[ -z "${LEASED_RESOURCE}" ]]; then
...
\ No newline at end of file
+if [[ -f "${SHARED_DIR}/terraform.tgz" ]]; then
+  mkdir -p /home/assisted-test-infra/build/terraform
+  cp ${SHARED_DIR}/terraform.tgz /home/assisted-test-infra/build/terraform
+  cd /home/assisted-test-infra/build/terraform
+  tar -xvf terraform.tgz
+  terraform init -input=false
+  terraform apply -destroy -input=false -auto-approve
+else
+  echo "Terraform file does not exist - probably websphere machine does not created - skipping teardown"
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/vsphere/teardown/machine/assisted-vsphere-teardown-machine-commands.sh:23]: ... probably websphere machine hasn't been created...

Comment (osherdp) [reply ci-operator/step-registry/assisted/vsphere/teardown/machine/assisted-vsphere-teardown-machine-commands.sh:23]: And let's short-circuit this code branch, meaning:
if <file does not exist>
  echo...
  exit 0

...

**Review by osherdp — CHANGES_REQUESTED** [review]

File: ci-operator/step-registry/assisted/vsphere/teardown/cluster/assisted-vsphere-teardown-cluster-commands.sh (hunk 8 above)
Comment (michaellevy101) [reply ci-operator/step-registry/assisted/vsphere/teardown/cluster/assisted-vsphere-teardown-cluster-commands.sh]: already deleted. just for debugging

File: ci-operator/step-registry/assisted/vsphere/teardown/cluster/assisted-vsphere-teardown-cluster-ref.yaml [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    requests:
+      cpu: 1000m
+      memory: 500Mi
+  credentials:
+  - namespace: test-credentials
+    name: vsphere-template-ssh-keys
+    mount_path: /var/run/vault/sshkeys
+  - namespace: test-credentials
+    name: edge-group-assisted-installer-bot
+    mount_path: /var/run/assisted-installer-bot
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/vsphere/teardown/cluster/assisted-vsphere-teardown-cluster-ref.yaml:15]: why do we need this?

File: ci-operator/step-registry/assisted/vsphere/teardown/machine/assisted-vsphere-teardown-machine-commands.sh [hunk 12]
```diff
@@ -12,9 +12,14 @@ if [[ -z "${LEASED_RESOURCE}" ]]; then
   exit 1
 fi
 
+if [[ ! -r "${SHARED_DIR}/terraform.tgz" ]]; then
+    echo "Terraform file does not exist - probably websphere machine does not created - skipping teardown"
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/vsphere/teardown/machine/assisted-vsphere-teardown-machine-commands.sh:16]: don't forget to fix grammar in this sentence

**Review by osherdp — CHANGES_REQUESTED** [review]

**Comment by osherdp** [comment]
/uncc
cc me back when needed

**Comment by michaellevy101** [comment]
@osherdp @gamli75 I added a cluster verification and everything is working as expected. please review and let's merge it as soon as possible

File: ci-operator/step-registry/assisted/common/verify/assisted-common-verify-commands.sh [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+echo "### Running tests"
+timeout --kill-after 10m 120m ssh -F ${SHARED_DIR}/ssh_config "root@ci_machine" bash - << EOF
+    # download openshift-tests cli tool from container quay.io/openshift/origin-tests
+    CONTAINER_ID=\$(podman run -d quay.io/openshift/origin-tests)
+    podman cp \${CONTAINER_ID}:/usr/bin/openshift-tests /usr/local/bin/
+    podman rm -f \${CONTAINER_ID}
+
+    # Update /etc/hosts
+    for FILE in /etc/NetworkManager/dnsmasq.d/*; do
+      awk '{split(\$0,a,"/"); print a[3] "\t\t" a[2] }' \${FILE} >> /etc/hosts
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/common/verify/assisted-common-verify-commands.sh:40]: why do we need this? we should still have DNS resolution for those, no?

File: ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+set -o errexit
+set -o pipefail
+
+echo "************ assisted common gather command ************"
+
+# TODO: Remove once OpenShift CI will be upgraded to 4.2 (see https://access.redhat.com/articles/4859371)
+~/fix_uid.sh
+
+cat > gather_logs.yaml <<-EOF
+- name: Gather logs and debug information and save them for debug proposes
```
Comment (osherdp) [inline ci-operator/step-registry/assisted/common/gather/assisted-common-gather-commands.sh:13]:
```suggestion
- name: Gather logs and debug information and save them for debug purposes
```

**Review by osherdp — CHANGES_REQUESTED** [review]

**Review by osherdp — APPROVED** [review]

**Comment by osherdp** [comment]
/hold
(waiting for rehearse)

**Pull request merged**